### ISSUE TITLE
Senior Attorney: Sign Stipulated Decision (1.8) 

### ISSUE THREAD

**Issue by JessicaMarine** [flexion/ef-cms#114]
As a Senior Attorney, in order to close a case, I want to sign the proposed Stipulated Decision submitted by the parties.

When a proposed Stipulated Decision is submitted to the Court by the parties, the document is routed to the Docket Section.  After reviewing the [Stipulated Decision,] (https://www.ustaxcourt.gov/taxpayer_info_glossary.htm#STIPULATED_DECISION) the Docket Clerk notifies a Senior Attorney that the Stipulated Decision is ready for final review and signature.  The Senior Attorney reviews the Stipulated Decision and, if everything is okay, needs to apply the signature of the Chief Judge.  After the Stipulated Decision is signed, it is sent back to the Docket Clerk for final processing.  The Docket Clerk creates a new docket entry for the *signed* Stipulated Decision.  The signed Stipulated Decision is viewable by all CMS users (members of the public, parties, internal court users).

## Test Scenarios:
https://docs.google.com/document/d/159FznIwhcWJlk2LYjI9yCDfASwCDgDZT6XFE5ZeV0U4/edit?usp=sharing 

## Pre-conditions:
* A proposed Stipulated Decision has been filed by the parties and has been manually routed to a Senior Attorney for review/signature -- shows up in Senior Attorney's individual and section work queue.
	
## Acceptance Criteria (Post-conditions):
* Senior Attorney can apply an electronic signature of the Chief Judge anywhere on the proposed Stipulated Decision.
* Senior Attorney can reapply the signature if they placed it incorrectly
* Senior Attorney can cancel signature process
* Senior Attorney can confirm signature
* The signed version of the Stipulated Decision is saved as a separate draft document. It does not overwrite the original.
* Senior Attorney MUST generate a message to the docket clerk during the signature process that appears on the Signed Stipulated Decision doc detail page, and this message appears in Senior Attorney's Sent Message box - 
* Once signed, the draft Stipulated Decision is viewable by all internal users, but not public or parties int he case. 
* Once signed, the message thread TO the Senior Attorney on the Proposed Stipulated Decision is closed and removed from their Message Inbox. 

## Notes
* Proposed Stipulated Decisions are of various sizes, so need the ability to apply the signature anywhere on the page.
* The digital signature will be a conformed signature of the judges signature
* Stipulated decisions are always 8.5x11; we don't need to scale the signature because of this. 
* Need the ability to drag and drop ("stamp")
* Need the ability to sign on any page
* Jessica said she envisions a "box" with transparent background
* PDF will be searchable. We'll have an original and then the signed copy.
* The UI is only a representation of the PDF, which gives us the coordinates of the signature
* On the back end it adds the signature to the doc
* Jessica - we would not want users typing it out every time. We'd want to control the content and who has access to the signature. 
* Managing the signature is easier with text than an image. 
* Jessica, for now, just used conformed signature (text). If it changes, we'll address it then. 

Which page the signature is likely to be on:
* 95% on the first page
* 4% on second
* 1% past that

## Tasks
- [x] POC - Drag / drop signature UI (Matt)
- [x] POC - Document generation with signature (Matt)
- [x] Document generation interactor (Matt)
- [x] Cycle for page selection, ability to drag signature after placed and ability to confirm and cancel (UX) (Sarah)
- [x] Implement cycle view (Matt ~*Done, but waiting on additional UX input~ #2199)
- [x] Service worker implementation (Cody)
- [x] Refactor drag and drop UI (Joe/Matt ~* Done, but waiting on lambda work for PR~ #2199)
- [x] Generate a document persistence; needs a lambda (Joe)
- [x] Testing scenarios (Mariah)
- [x] Apply updated design to signature view (Matt #2210)
- [x] Dispatch message to docket clerk (Joe #2219)
- [x] Complete Proposed Stip Decision message (Joe #2211)
- [x] Add opaque white box to signature background on PDF (Joe #2233)
- [x] Add opaque white box to signature background on view
- [x] Update design based on 7/1 feedback (Sarah, Matt #2227)
- [x] Refactoring and clean up (Cody/Matt/Joe #2258)
- [x] Add user role/title to signature on view and PDF (Joe #2241)
- [x] Work queue flow issues (Cody/Matt)

 ## Definition of Done
**Product Owner**
 - [ ] Acceptance criteria have been met
**UX**
  - [x] Usability has been validated
**Engineering**
  - [x] Field level and page level validation errors (front-end and server side) integrated and functioning
  - [x] ~Run the browser stack scripts to verify that application is usable in Chrome, Firefox, Safari, Edge and Internet Explorer (external only)~ N/A 
  - [x] No errors reported in manual a11y testing (Eric)
  - [x] ~All new functionality verified to work with keyboard and osx voiceover~ N/A
  - [x] Code refactored for clarity and to remove any known technical debt (Joe)
  - [x]  README, other appropriate docs, and swagger/APIs fully updated - all code functions documented with jsdocs (Cody)
  - [x] ~UI should be touch optimized and responsive for external only(functions on supported mobile devices and optimized for screen sizes as required)~ N/A
  - [x] Library/module dependencies are up-to-date and are at the latest resolvable version (Eric)
  - [x] Code is deployed to DEV
  - [x] Errors in sonarcloud are fixed (https://sonarcloud.io/organizations/flexion-github/projects) (Joe)
  - [x] Alerting and monitoring for new functionality in place and successfully exercised (Matt)

**Comment by klohman** [comment]
Design for Senior Attorney Work Queue (reviewed 12/31 with Jessica as part of 406)

 ![Senior Attorney Queue.png](https://images.zenhubusercontent.com/5bd7271508843955cc24d3d0/2e550557-db1d-4348-bb08-10939ca89adf)

**Comment by mariahkannenberg** [comment]
Story for dealing with draft stipulated decision when it's incorrect added #2184

**Comment by mariahkannenberg** [comment]
Test Scenarios - Sign Stipulated Decision

*Scenario 1: Sr. Attorney signs Proposed Stipulated Decision (DRAFT)*

Practitioner1 logs into the system
Practitioner1 navigates to case
Practitioner1 selects File a Document
Practitioner1 selects “Decisions” from Document Category dropdown
Practitioner1 selects “Proposed Stipulated Decision” from Document Type dropdown
Practitioner1 selects “Select” button
Practitioner1 selects “Continue” button
Practitioner1 selects “Choose File” button and uploads proposed stipulated decision
Practitioner1 selects “No” to question “Does your file include a certificate of service?”
Practitioner1 selects “No” to question “Does your filing include exhibits?”
Practitioner1 selects “No” to question “Does your filing include attachments?”
Practitioner1 selects “No” to question “Do you have any supporting documents for this filing?”
Practitioner1 selects “Myself as Petitioner’s Counsel” to question “Who is filing this document.”
Practitioner1 selects “Review FIling” button
Practitioner1 selects “Submit Your Filing” button
Practitioner1 logs out
Docketclerk1 logs in
Docketclerk1 navigates to Document QC>Section Documents
Docketclerk1 selects Proposed Stipulated Decision
Docketclerk1 selects Create Message
Docketclerk1 selects Senior Attorney from Select Section dropdown
Docketclerk1 selects Test seniorattorney1 from the Select Recipient dropdown
Docketclerk1 types “Jeff, this is ready for review and signature” into the Add Message field
Docketclerk1 selects Send
Docketclerk1 logs out
Seniorattorney1 logs in
Seniorattorney1 selects Proposed Stipulated Decision message sent by docketclerk1
Seniorattorney1 selects Add Signature Button
Seniorattorney1 selects Apply Signature Button
Seniorattorney1 hovers over document and clicks to place signature on the document
Seniorattorney1 selects Docket from Select Section dropdown
Seniorattorney1 selects docketclerk1 from Select Recipient dropdown
Seniorattorney1 types “Donna, this is not ready to serve. I need to follow up on something first” in the Add Messages field
Seniorattorney1 selects Save (Is there a step to send the message? Does it happen with ‘Save’)
Seniorattorney1 is auto-directed to (? my messages?) after finishing the signature process
Seniorattorney1 does not see the Proposed Stipulated Decision message from the docket clerk in the inbox (Do we need to auto-remove it? Need to get rid of it in some way, right?)
Seniorattorney1 navigates to sent box and sees Signed Stipulated Decision
Seniorattorney1 selects Signed Stipulated Decision and sees the message they generated to docketclerk1 during the signature process “Donna, this is not ready to serve. I need to follow up on something first”
SeniorAttorney1 logs out
Docketclerk1 logs in
Docketclerk1 navigates to My Messages>Sent and sees the Proposed Stipulated Decision
Docketclerk1 selects Proposed Stipulated Decision, then selects Complete and sees that the message to senior has been completed (automatically when sr attorney applies signature) “Jeff, this is ready for review and signature”
Docketclerk1 navigates to My Messages>Inbox and sees Signed Stipulated Decision message from Sr. Attorney “Donna, this is not ready to serve. I need to follow up on something first” 
Docketclerk1 select Signed Stipulated Decision and sees the document in draft form
Docketclerk1 logs out


*Scenario 2: Sr. Attorney views signed stip decision in “draft queue”*



*Scenario 3: Docket Clerk views signed stip decision in “draft queue”*


*Scenario 4: Judge Signs Stip Decision Uploaded by Petitioner (Don’t yet have judge’s dashboard to test. Test once built.)*
Practitioner1 logs into the system
Practitioner1 navigates to case
Practitioner1 selects File a Document
Practitioner1 selects “Decisions” from Document Category dropdown
Practitioner1 selects “Proposed Stipulated Decision” from Document Type dropdown
Practitioner1 selects “Select” button
Practitioner1 selects “Continue” button
Practitioner1 selects “Choose File” button and uploads proposed stipulated decision
Practitioner1r selects “No” to question “Does your file include a certificate of service?”
Practitioner1 selects “No” to question “Does your filing include exhibits?”
Practitioner1 selects “No” to question “Does your filing include attachments?”
Practitioner1 selects “No” to question “Do you have any supporting documents for this filing?”
Practitioner1 selects “Myself as Petitioner’s Counsel” to question “Who is filing this document.”
Practitioner1 selects “Review FIling” button
Practitioner1 selects “Submit Your Filing” button
Practitioner1 logs out
Judge logs in
Judge navigates to Document QC for Section
Judge clicks on newly uploaded Proposed Stipulated Decision
Judge signs proposed stipulated decision with judge’s conformed signature (NEW UI, Design TBD)
After applying signature, Judge is automatically directed to document detail page of document that has signature
Judge selects “Create a Message”
Judge selects “Docket” from Select Section dropdown
Judge selects “Test docketclerk1” from Select Recipient dropdown
Judge types “The stipulated decision has been signed and is ready to serve” into the add message field
Judge selects “Send”
Judge logs out
Docketclerk1 logs in and sees new unread message for Proposed Stipulated Decision
Docketclerk1 selects “Proposed Stipulated Decision”
Docketclerk1 selects Docket Number
Docket clerks selects “Proposed Stipulated Decision” in the docket record and sees the unsigned version of the decision
Docketclerk1 selects Docket Number
Docket clerk selects “Signed Proposed Stipulated Decision” in the docket record and sees a PDF of the signed stipulated decision (NEW UI, Design TBD)
Docketclerk1 logs out
Practitioner1 logs in
Practitioner1 selects case for which they submitted a proposed stip decision
Practitioner1 selects Docket Number
Practitioner1 selects “Proposed Stipulated Decision” in the docket record and sees the unsigned version of the decision
Practitioner1 selects Docket Number
Practitioner1 selects “Signed Proposed Stipulated Decision” in the docket record and sees a PDF of the signed stipulated decision (NEW UI, Design TBD)
Practitioner1 logs out

**Comment by sarahridgway** [comment]
*User Flow*

![image.png](https://images.zenhubusercontent.com/5be0bad045b0290c78fb7458/abcbdd4a-83ef-4844-a1a9-d86d4e0f2e91)

**Comment by sarahridgway** [comment]
*Proposed Stipulated Decision - Document Detail Page*

- Senior Attorney clicks on message in My Message Inbox, opens Document Detail Page
- Clicks on "Sign Document" button, opens Sign Document Page

![image.png](https://images.zenhubusercontent.com/5be0bad045b0290c78fb7458/45e0ccc6-4ec6-457f-be1d-78f56f0ab8b1)

*Proposed Stipulated Decision - Sign Document Page - Not Signed*

- Use page template for Order Generation (Themis) - PDF canvas [right] and tool bar [left]
- Use page navigation to flip through pages
- Click "Apply Signature" button to begin placing Chief Judge's signature
- "Save" button is disabled until a signature has been placed on the PDF canvas AND a message has been created with a recipient selected
-  Click "Cancel" link to cancel and return to the Proposed Stipulated Decision Document Detail Page - confirmation modal is not necessary
- Use Create Message template from Doc Detail - Create Message Modal with "(required)" note

![image.png](https://images.zenhubusercontent.com/5be0bad045b0290c78fb7458/446a707f-3f7f-442c-a7fe-d01a6d3b3739)

*Proposed Stipulated Decision - Sign Document Page - Signed*

- "Apply Signature" button attaches the conformed (plain text) signature to the user's cursor
- Disable "Apply Signature" button when the signature is attached to cursor or canvas

- Use 'grab' cursor when signature is attached to cursor
- Place signature when the user clicks on the canvas, co-ordinates are obtained

- "Clear Signature" link appears when the "Apply Signature" button has been clicked
- "Clear Signature" link allows the user to remove a placed signature from the canvas
- When "Clear Signature" link is clicked, re-activate "Apply Signature" button, remove clear link

- Signature on canvas is required
- Message and recipient are required
- When signature and message are complete, active "Save" button

- Saving will complete / close the message on the Proposed Stipulated Decision Doc Detail
- Proposed Stipulated Decision will remain in docket record
- Saving will generate a PDF of the Signed Stipulated Decision
- Signed Stipulated Decision PDF will be placed in a Draft Document tab on the Case Detail (possibly a separate story)
- Message will appear on Signed Stipulated Decision Doc Detail and be sent to the selected recipient's Message Inbox as well as the Senior Attorney's Messages Sent box.
- After user has Saved document, send user to Case Detail Page - Draft Document tab (if draft document tab is moved to a separate story, send user to their Message Inbox)

![image.png](https://images.zenhubusercontent.com/5be0bad045b0290c78fb7458/13b40cdc-a91f-4c02-92d4-de02f2fb4206)

*Draft Documents - Case Detail Page*

Note: This may be moved to a separate story

- Signed Stipulated Decision appears in the new Draft Documents tab
- Clicking the Document Link opens Document Detail Page
- Documents in the Draft Documents tab are not shown on the Docket Record and are not accessible to external users
- If Signed Stipulated Decision PDF generation is still processing, show the processing spinner in the document row

*Proposed Stipulated Decision - Document Detail Page - Complete* 

- When Signed Stipulated Decision has been saved, remote "Sign Document" button from Proposed Stipulated Decision - Document Detail Page

![image.png](https://images.zenhubusercontent.com/5be0bad045b0290c78fb7458/6e85defe-6039-47c5-a5ca-1df0f5f8f5a0)

**Comment by sarahridgway** [comment]
@JessicaMarine How should the signature appear on the PDF? (Font / Size / Color)

**Comment by sarahridgway** [comment]
@mariahkannenberg 

Re: test scenarios above

Seniorattorney1 selects Save (Is there a step to send the message? Does it happen with ‘Save’)

- Message is sent when the user clicks "Save"

Seniorattorney1 is auto-directed to (? my messages?) after finishing the signature process

- If draft documents tab is implemented in this story, send user to the draft documents tab so they can see the Signed Stipulated Decision has been created. If not, send user to their message inbox. Do we need a success message?

Seniorattorney1 does not see the Proposed Stipulated Decision message from the docket clerk in the inbox (Do we need to auto-remove it? Need to get rid of it in some way, right?)

- Complete message automatically upon Saving the signature, message drops off the inbox.

**Comment by JessicaMarine** [comment]
Arial/Size 14/Black

Sample:
(signed) David Gustafson
Judge

**Comment by JessicaMarine** [comment]
@sarahridgway 

Arial Bold/Size 14/Black

Sample:
**(signed) Maurice B. Foley
             Chief Judge**

**Comment by sarahridgway** [comment]
Does the signature include "Chief Judge" or would that be part of the actual document, indicating where the Judge's signature should go?

**Comment by JessicaMarine** [comment]
PO Review 7/10/19:
* After signature is applied and saved, the new PDF that is generated should drop the cover page (I missed this in the AC, so I can add it to a clean-up story if it's too big to fix here)
* The signature box should not have a black line as an outline around it
* The signature should be static with a Judge's name (here "Maurice B. Foley"), not dynamically insert the user's name
* There is no error handling on the send message function  after signing -- I didn't put anyone's info in and hit Save & Send and it seemed to do something as it routed me back to my Doc Inbox.
* No docket entry should be created when the new PDF is created -- it should be saved somewhere in the system as a draft document that is accessible from the message thread (i.e., Senior Atty's Outbox and Docket Clerk Inbox), but not viewable on the docket record yet
*  If someone goes into that draft document from the message thread, the title of that page should be "Signed Stipulated Decision" -- it's still showing as "Proposed Stipulated Decision"
